File: machine/ts/machine/devices/audio/save_state.ts
import type { ApuAudioSlot } from './contracts';
import type { BiquadFilterState } from './biquad_filter';
import type { ApuCommandFifoState } from './command_fifo';

export type ApuSampleTransferState = {
	transferAddressWord: number;
	transferDataWord: number;
	transferControlWord: number;
	currentAddress: number;
	fifoWords: number[];
	fifoReadIndex: number;
	fifoWriteIndex: number;
	fifoCount: number;
	timingCarry: number;
	scheduledWords: number;
	scheduledCycles: number;
};

export type ApuBiquadFilterState = {
	l1: number;
	l2: number;
	r1: number;
	r2: number;
};

export type ApuBadpDecoderSaveState = {
	predictors: number[];
	stepIndices: number[];
	nextFrame: number;
	blockEnd: number;
	blockFrames: number;
	blockFrameIndex: number;
	payloadOffset: number;
	nibbleCursor: number;
	decodedFrame: number;
	decodedLeft: number;
	decodedRight: number;
	previousDecodedFrame: number;
	previousDecodedLeft: number;
	previousDecodedRight: number;
};

export type ApuOutputVoiceState = {
	slot: ApuAudioSlot;
	cursorQ16: number;
	phaseRemainder: number;
	gainQ12: number;
	fadeStepQ12: number;
	fadeStepRemainder: number;
	fadeError: number;
	fadeSamplesRemaining: number;
	fadeSamplesTotal: number;
	filter: ApuBiquadFilterState;
	badp: ApuBadpDecoderSaveState;
};

export type ApuOutputState = {
	voices: ApuOutputVoiceState[];
};

export type AudioControllerState = {
	registerWords: number[];
	commandFifo: ApuCommandFifoState;
	eventSequence: number;
	eventKind: number;
	eventSlot: number;
	eventSourceAddr: number;
	slotPhases: number[];
	slotRegisterWords: number[];
	sampleRam: Uint8Array;
	sampleTransfer: ApuSampleTransferState;
	output: ApuOutputState;
	sampleCarry: number;
	sampleSequence: number;
	apuStatus: number;
	apuFaultCode: number;
	apuFaultDetail: number;
};

export type ApuMutableNumberArrayLike = ArrayLike<number> & { [index: number]: number };

export type ApuBadpDecoderStateAccess = Omit<ApuBadpDecoderSaveState, 'predictors' | 'stepIndices'> & {
	predictors: ApuMutableNumberArrayLike;
	stepIndices: ApuMutableNumberArrayLike;
};

export type ApuOutputVoiceStateAccess = Omit<ApuOutputVoiceState, 'filter' | 'badp'> & {
	filter: BiquadFilterState;
	badp: ApuBadpDecoderStateAccess;
};

export function captureApuOutputVoiceState(record: ApuOutputVoiceStateAccess): ApuOutputVoiceState {
	return {
		slot: record.slot,
		cursorQ16: record.cursorQ16,
		phaseRemainder: record.phaseRemainder,
		gainQ12: record.gainQ12,
		fadeStepQ12: record.fadeStepQ12,
		fadeStepRemainder: record.fadeStepRemainder,
		fadeError: record.fadeError,
		fadeSamplesRemaining: record.fadeSamplesRemaining,
		fadeSamplesTotal: record.fadeSamplesTotal,
		filter: {
			l1: record.filter.l1,
			l2: record.filter.l2,
			r1: record.filter.r1,
			r2: record.filter.r2,
		},
		badp: {
			predictors: Array.from(record.badp.predictors),
			stepIndices: Array.from(record.badp.stepIndices),
			nextFrame: record.badp.nextFrame,
			blockEnd: record.badp.blockEnd,
			blockFrames: record.badp.blockFrames,
			blockFrameIndex: record.badp.blockFrameIndex,
			payloadOffset: record.badp.payloadOffset,
			nibbleCursor: record.badp.nibbleCursor,
			decodedFrame: record.badp.decodedFrame,
			decodedLeft: record.badp.decodedLeft,
			decodedRight: record.badp.decodedRight,
			previousDecodedFrame: record.badp.previousDecodedFrame,
			previousDecodedLeft: record.badp.previousDecodedLeft,
			previousDecodedRight: record.badp.previousDecodedRight,
		},
	};
}

export function restoreApuOutputVoiceState(record: ApuOutputVoiceStateAccess, state: ApuOutputVoiceState): void {
	record.cursorQ16 = state.cursorQ16;
	record.phaseRemainder = state.phaseRemainder;
	record.gainQ12 = state.gainQ12;
	record.fadeStepQ12 = state.fadeStepQ12;
	record.fadeStepRemainder = state.fadeStepRemainder;
	record.fadeError = state.fadeError;
	record.fadeSamplesRemaining = state.fadeSamplesRemaining;
	record.fadeSamplesTotal = state.fadeSamplesTotal;
	record.filter.l1 = state.filter.l1;
	record.filter.l2 = state.filter.l2;
	record.filter.r1 = state.filter.r1;
	record.filter.r2 = state.filter.r2;
	record.badp.predictors[0] = state.badp.predictors[0]!;
	record.badp.predictors[1] = state.badp.predictors[1]!;
	record.badp.stepIndices[0] = state.badp.stepIndices[0]!;
	record.badp.stepIndices[1] = state.badp.stepIndices[1]!;
	record.badp.nextFrame = state.badp.nextFrame;
	record.badp.blockEnd = state.badp.blockEnd;
	record.badp.blockFrames = state.badp.blockFrames;
	record.badp.blockFrameIndex = state.badp.blockFrameIndex;
	record.badp.payloadOffset = state.badp.payloadOffset;
	record.badp.nibbleCursor = state.badp.nibbleCursor;
	record.badp.decodedFrame = state.badp.decodedFrame;
	record.badp.decodedLeft = state.badp.decodedLeft;
	record.badp.decodedRight = state.badp.decodedRight;
	record.badp.previousDecodedFrame = state.badp.previousDecodedFrame;
	record.badp.previousDecodedLeft = state.badp.previousDecodedLeft;
	record.badp.previousDecodedRight = state.badp.previousDecodedRight;
}

File: machine/cpp/machine/devices/audio/save_state.cpp
#include "machine/devices/audio/controller.h"
#include "machine/devices/audio/output.h"

#include "machine/bus/io.h"
#include "machine/cpu/value.h"
#include "machine/scheduler/device.h"

namespace bmsx {

ApuOutputVoiceState captureApuOutputVoiceState(const ApuOutputMixer::VoiceRecord& record) {
	ApuOutputVoiceState voice;
	voice.slot = record.slot;
	voice.sourceCartridgeSlot = record.sourceCartridgeSlot;
	voice.cursorQ16 = record.cursorQ16;
	voice.phaseRemainder = record.phaseRemainder;
	voice.gainQ12 = record.gainQ12;
	voice.fadeStepQ12 = record.fadeStepQ12;
	voice.fadeStepRemainder = record.fadeStepRemainder;
	voice.fadeError = record.fadeError;
	voice.fadeSamplesRemaining = record.fadeSamplesRemaining;
	voice.fadeSamplesTotal = record.fadeSamplesTotal;
	voice.filter.l1 = record.filter.l1;
	voice.filter.l2 = record.filter.l2;
	voice.filter.r1 = record.filter.r1;
	voice.filter.r2 = record.filter.r2;
	voice.badp.predictors[0] = record.badp.predictors[0];
	voice.badp.predictors[1] = record.badp.predictors[1];
	voice.badp.stepIndices[0] = record.badp.stepIndices[0];
	voice.badp.stepIndices[1] = record.badp.stepIndices[1];
	voice.badp.nextFrame = static_cast<u32>(record.badp.nextFrame);
	voice.badp.blockEnd = static_cast<u32>(record.badp.blockEnd);
	voice.badp.blockFrames = static_cast<u32>(record.badp.blockFrames);
	voice.badp.blockFrameIndex = static_cast<u32>(record.badp.blockFrameIndex);
	voice.badp.payloadOffset = static_cast<u32>(record.badp.payloadOffset);
	voice.badp.nibbleCursor = static_cast<u32>(record.badp.nibbleCursor);
	voice.badp.decodedFrame = record.badp.decodedFrame;
	voice.badp.decodedLeft = record.badp.decodedLeft;
	voice.badp.decodedRight = record.badp.decodedRight;
	voice.badp.previousDecodedFrame = record.badp.previousDecodedFrame;
	voice.badp.previousDecodedLeft = record.badp.previousDecodedLeft;
	voice.badp.previousDecodedRight = record.badp.previousDecodedRight;
	return voice;
}

void restoreApuOutputVoiceState(ApuOutputMixer::VoiceRecord& record, const ApuOutputVoiceState& state) {
	record.sourceCartridgeSlot = state.sourceCartridgeSlot;
	record.cursorQ16 = state.cursorQ16;
	record.phaseRemainder = state.phaseRemainder;
	record.gainQ12 = state.gainQ12;
	record.fadeStepQ12 = state.fadeStepQ12;
	record.fadeStepRemainder = state.fadeStepRemainder;
	record.fadeError = state.fadeError;
	record.fadeSamplesRemaining = state.fadeSamplesRemaining;
	record.fadeSamplesTotal = state.fadeSamplesTotal;
	record.filter.l1 = state.filter.l1;
	record.filter.l2 = state.filter.l2;
	record.filter.r1 = state.filter.r1;
	record.filter.r2 = state.filter.r2;
	record.badp.predictors[0] = state.badp.predictors[0];
	record.badp.predictors[1] = state.badp.predictors[1];
	record.badp.stepIndices[0] = state.badp.stepIndices[0];
	record.badp.stepIndices[1] = state.badp.stepIndices[1];
	record.badp.nextFrame = state.badp.nextFrame;
	record.badp.blockEnd = state.badp.blockEnd;
	record.badp.blockFrames = state.badp.blockFrames;
	record.badp.blockFrameIndex = state.badp.blockFrameIndex;
	record.badp.payloadOffset = state.badp.payloadOffset;
	record.badp.nibbleCursor = state.badp.nibbleCursor;
	record.badp.decodedFrame = state.badp.decodedFrame;
	record.badp.decodedLeft = state.badp.decodedLeft;
	record.badp.decodedRight = state.badp.decodedRight;
	record.badp.previousDecodedFrame = state.badp.previousDecodedFrame;
	record.badp.previousDecodedLeft = state.badp.previousDecodedLeft;
	record.badp.previousDecodedRight = state.badp.previousDecodedRight;
}

AudioControllerState AudioController::captureState() {
	const i64 nowCycles = m_scheduler.currentNowCycles();
	m_serviceClock.synchronize(nowCycles);
	AudioControllerState state;
	for (size_t index = 0; index < APU_PARAMETER_REGISTER_COUNT; index += 1u) {
		state.registerWords[index] = m_memory.readIoU32(IO_APU_PARAMETER_REGISTER_ADDRS[index]);
	}
	state.commandFifo = m_commandFifo.captureState();
	const ApuEventLatchState event = m_eventLatch.captureState();
	state.eventSequence = event.eventSequence;
	state.eventKind = event.eventKind;
	state.eventSlot = event.eventSlot;
	state.eventSourceAddr = event.eventSourceAddr;
	state.slotPhases = m_slots.slotPhases();
	state.slotRegisterWords = m_slots.slotRegisterWords();
	state.sampleRam = m_sampleMemory.captureState();
	state.sampleTransfer = m_serviceClock.captureSampleTransferState(nowCycles);
	state.output = m_audioOutput.captureState();
	state.sampleCarry = m_serviceClock.captureSampleCarry();
	state.sampleSequence = m_serviceClock.captureSampleSequence();
	state.apuStatus = m_fault.status;
	state.apuFaultCode = m_fault.code;
	state.apuFaultDetail = m_fault.detail;
	return state;
}

void AudioController::restoreState(const AudioControllerState& state, int64_t nowCycles) {
	m_audioOutput.resetPlaybackState();
	for (size_t index = 0; index < APU_PARAMETER_REGISTER_COUNT; index += 1u) {
		m_memory.writeIoValue(IO_APU_PARAMETER_REGISTER_ADDRS[index], valueNumber(static_cast<double>(state.registerWords[index])));
	}
	m_commandFifo.restoreState(state.commandFifo);
	m_eventLatch.restoreState({state.eventSequence, state.eventKind, state.eventSlot, state.eventSourceAddr});
	m_slots.restore(state.slotPhases, state.slotRegisterWords);
	m_sampleMemory.restoreState(state.sampleRam);
	m_serviceClock.restore(state.sampleCarry, state.sampleSequence, state.sampleTransfer, nowCycles);
	m_fault.restore(state.apuStatus, state.apuFaultCode, state.apuFaultDetail);
	m_activeSlots.writeActiveMask();
	for (const ApuOutputVoiceState& voiceState : state.output.voices) {
		m_commandExecutor.restoreOutputVoice(voiceState);
	}
	m_serviceClock.scheduleNext(nowCycles);
}

ApuOutputState ApuOutputMixer::captureState() const {
	ApuOutputState state;
	state.voices.reserve(APU_SLOT_COUNT);
	for (const VoiceRecord& record : m_voices) {
		if (record.active) {
			state.voices.push_back(captureApuOutputVoiceState(record));
		}
	}
	return state;
}

} // namespace bmsx
